### PR TITLE
Optimize array element function when input is array(map)/array(array(num)/array(array(string))/array(bigint)/array(decimal)

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimize array element function when input is array(map)/array(array(num)/array(array(string))/array(bigint)/array(decimal). Current implementation causes too many reallocs. The optimization speed up by ~6x especially when input type is array(map)

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 7edc2542adbb17b9100e91c1531b8b19297946d8 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56403/7edc2542adbb17b9100e91c1531b8b19297946d8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6781803044">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56403/7edc2542adbb17b9100e91c1531b8b19297946d8/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by taiyang-li** [comment]
Some tests

# Optimize when input type is array(array(num))
Q: `select materialize(array(array(1,2,3,4)))[1] from numbers(10000000) format Null;`

Before: 
```
0 rows in set. Elapsed: 0.551 sec. Processed 10.01 million rows, 80.06 MB (18.18 million rows/s., 145.41 MB/s.)
Peak memory usage: 4.26 MiB.
0 rows in set. Elapsed: 0.528 sec. Processed 10.01 million rows, 80.06 MB (18.95 million rows/s., 151.63 MB/s.)
Peak memory usage: 4.26 MiB.
0 rows in set. Elapsed: 0.566 sec. Processed 10.01 million rows, 80.06 MB (17.67 million rows/s., 141.35 MB/s.)
Peak memory usage: 4.25 MiB.
```

After
```
0 rows in set. Elapsed: 0.464 sec. Processed 10.01 million rows, 80.06 MB (21.55 million rows/s., 172.41 MB/s.)
Peak memory usage: 4.25 MiB.
0 rows in set. Elapsed: 0.455 sec. Processed 10.01 million rows, 80.06 MB (21.99 million rows/s., 175.93 MB/s.)
Peak memory usage: 4.25 MiB.
0 rows in set. Elapsed: 0.470 sec. Processed 10.01 million rows, 80.06 MB (21.28 million rows/s., 170.22 MB/s.)
Peak memory usage: 4.25 MiB
```

# Optimize when input type is array(array(string))

Q: `select materialize(array(array('hello', 'world')))[1] from numbers(10000000)  format Null `

Before:
```
0 rows in set. Elapsed: 0.824 sec. Processed 10.01 million rows, 80.06 MB (12.14 million rows/s., 97.16 MB/s.)
Peak memory usage: 12.07 MiB.
0 rows in set. Elapsed: 0.827 sec. Processed 10.01 million rows, 80.06 MB (12.10 million rows/s., 96.78 MB/s.)
Peak memory usage: 12.07 MiB.
0 rows in set. Elapsed: 0.842 sec. Processed 10.01 million rows, 80.06 MB (11.88 million rows/s., 95.06 MB/s.)
Peak memory usage: 12.07 MiB. 
```

After:
```
0 rows in set. Elapsed: 0.722 sec. Processed 10.01 million rows, 80.06 MB (13.87 million rows/s., 110.94 MB/s.)
Peak memory usage: 12.06 MiB.
0 rows in set. Elapsed: 0.712 sec. Processed 10.01 million rows, 80.06 MB (14.05 million rows/s., 112.37 MB/s.)
Peak memory usage: 12.06 MiB.
0 rows in set. Elapsed: 0.722 sec. Processed 10.01 million rows, 80.06 MB (13.86 million rows/s., 110.86 MB/s.)
Peak memory usage: 12.06 MiB.
```

# Optimize when input type is map with its key/value type string/num
Q:  `select materialize(array(map('hello', 1, 'world', 2)))[1] from numbers(10000000) format Null;`

Before: 
```
0 rows in set. Elapsed: 6.504 sec. Processed 10.01 million rows, 80.06 MB (1.54 million rows/s., 12.31 MB/s.)
Peak memory usage: 12.58 MiB.
0 rows in set. Elapsed: 6.394 sec. Processed 10.01 million rows, 80.06 MB (1.57 million rows/s., 12.52 MB/s.)
Peak memory usage: 12.58 MiB.
0 rows in set. Elapsed: 6.398 sec. Processed 10.01 million rows, 80.06 MB (1.56 million rows/s., 12.51 MB/s.)
Peak memory usage: 12.58 MiB.
```

After: 
```
0 rows in set. Elapsed: 1.029 sec. Processed 10.01 million rows, 80.06 MB (9.72 million rows/s., 77.79 MB/s.)
Peak memory usage: 12.95 MiB.
0 rows in set. Elapsed: 1.010 sec. Processed 10.01 million rows, 80.06 MB (9.91 million rows/s., 79.26 MB/s.)
Peak memory usage: 12.95 MiB.
0 rows in set. Elapsed: 1.032 sec. Processed 10.01 million rows, 80.06 MB (9.70 million rows/s., 77.58 MB/s.)
Peak memory usage: 12.95 MiB.
```

**Comment by taiyang-li** [comment]
@alexey-milovidov  can you help review it, thanks very much!

**Comment by taiyang-li** [comment]
@Avogar can you help review it, thanks very much !

**Comment by alexey-milovidov** [comment]
It looks very promising; I didn't have time to review it yet.

**Comment by taiyang-li** [comment]
Do we have any progress recently ?

**Comment by taiyang-li** [comment]
Do we have any progress recently ?

**Comment by Avogar** [comment]
I will review this PR this week. Sorry for long wait.

File: src/Functions/array/arrayElement.cpp [hunk 1]
```diff
@@ -256,6 +301,363 @@ struct ArrayElementNumImpl
...
+                    builder.update(j);
+
+                ColumnArray::Offset nested_array_size = nested_offsets[j] - nested_offsets[j-1];
+                ColumnArray::Offset nested_array_pos = nested_offsets[j-1];
+                for (size_t k = 0; k < nested_array_size; ++k)
+                {
+                    result_data[current_offset + k] = data[nested_array_pos + k];
+                    if constexpr (nullable_number)
+                        (*result_number_null_map)[current_offset + k] = (*number_null_map)[nested_array_pos + k];
+                }
```
Comment (Avogar) [inline src/Functions/array/arrayElement.cpp:363]: Maybe here and in all similar places we can just do memcpy instead of a loop

**Comment by taiyang-li** [comment]
> Actually, I think we can generalize new implementations a bit to support any nested arrays, not only 2 level arrays. We will just need to use a vector of nested and result offsets. Not sure if there will be an overhead for 2 level arrays because of iterating through vector with single element. WDYT? Current code LGTM, only one small comment.
> 
> Also need to make new perf tests faster, they are too slow: [https://s3.amazonaws.com/clickhouse-test-reports/56403/4bbc0db1540bae020fecc1f1d8ff3bcc4eb191c2/performance_comparison_[1_4]/report.html](https://s3.amazonaws.com/clickhouse-test-reports/56403/4bbc0db1540bae020fecc1f1d8ff3bcc4eb191c2/performance_comparison_%5B1_4%5D/report.html)

I'll commit again, and refersh the perf tests(current result seems not precise, similar situations happended in other prs before). You thoughts about support any nested arrsys is interesting, but I'm consused about how to implement it, do you have any example.

**Comment by Avogar** [comment]
> You thoughts about support any nested arrsys is interesting, but I'm consused about how to implement it, do you have any example.

I started to write an explanation, but it turned out to be difficult to explain without just writing this code :)
Let's merge current implementation, it's also really good. I will try to implement optimization for arrays with any level by myself.

**Comment by Avogar** [comment]
Perf test still complains about `array_element` tests, they are too long, we should make them run under 1 second

File: tests/performance/array_element.xml [hunk 2]
```diff
@@ -2,4 +2,20 @@
...
+    <query>select materialize(array(array('hello', 'world')))[1] from numbers(10000000)</query>
+    <query>select materialize(array(array('hello', 'world')))[materialize(1)] from numbers(10000000)</query>
+
+    <!-- arrayElement function with input type Array(Array(Map(String, UInt8))) -->
+    <query>select materialize(array(map('hello', 1, 'world', 2)))[1] from numbers(10000000)</query>
+    <query>select materialize(array(map('hello', 1, 'world', 2)))[materialize(1)] from numbers(10000000)</query>
+
+    <!-- arrayElement function with input type Array(Decimal256(2)) -->
+    <query>select materialize(array(1.23::Decimal256(2), 4.56::Decimal256(2)))[1] from numbers(10000000)</query>
+    <query>select materialize(array(1.23::Decimal256(2), 4.56::Decimal256(2)))[materialize(1)] from numbers(10000000)</query>
```
Comment (Avogar) [inline tests/performance/array_element.xml:20]:
```suggestion
    <query>select materialize(array(array(1,2,3,4)))[1] from numbers(10000000) fromat Null</query>
    <query>select materialize(array(array(1,2,3,4)))[materialize(1)] from numbers(10000000) format Null</query>

    <!-- arrayElement function with input type Array(Array(String)) -->
    <query>select materialize(array(array('hello', 'world')))[1] from numbers(10000000) format Null</query>
    <query>select materialize(array(array('hello', 'world')))[materialize(1)] from numbers(10000000) format Null</query>

    <!-- arrayElement function with input type Array(Array(Map(String, UInt8))) -->
    <query>select materialize(array(map('hello', 1, 'world', 2)))[1] from numbers(10000000) format Null</query>
    <query>select materialize(array(map('hello', 1, 'world', 2)))[materialize(1)] from numbers(10000000) format Null</query>

    <!-- arrayElement function with input type Array(Decimal256(2)) -->
    <query>select materialize(array(1.23::Decimal256(2), 4.56::Decimal256(2)))[1] from numbers(10000000) format Null</query>
    <query>select materialize(array(1.23::Decimal256(2), 4.56::Decimal256(2)))[materialize(1)] from numbers(10000000) format Null</query>
```

**Comment by taiyang-li** [comment]
image

File: tests/performance/array_element.xml [hunk 3]
```diff
@@ -2,4 +2,20 @@
     <query>SELECT count() FROM numbers(100000000) WHERE NOT ignore([[1], [2]][number % 2 + 2])</query>
     <query>SELECT count() FROM numbers(100000000) WHERE NOT ignore([[], [2]][number % 2 + 2])</query>
     <query>SELECT count() FROM numbers(100000000) WHERE NOT ignore([[], []][number % 2 + 2])</query>
+
+    <!-- arrayElement function with input type Array(Array(UInt8)) -->
+    <query>select materialize(array(array(1,2,3,4)))[1] from numbers(10000000) fromat Null</query>
```
Comment (Avogar) [inline tests/performance/array_element.xml:7]:
```suggestion
    <query>select materialize(array(array(1,2,3,4)))[1] from numbers(10000000) format Null</query>
```
My bad, mad a typo in previous suggestion

**Comment by taiyang-li** [comment]
@Avogar  I don't known why there are always two errors about `parallel_index` in perf tests. I'll merge master and observe it again.

**Comment by Avogar** [comment]
Error in `parallel_index` - https://github.com/ClickHouse/ClickHouse/issues/58593

**Review by Avogar — APPROVED** [review]

**Pull request merged**